Reset `DataLoader` workers instead of creating new ones

emcastillo opened the pull request:
This PR needs discussion as it changes the behavior of `DataLoader`. It can be closed if its not considered a good practice.

Currently, the `DataLoader` spawns a new `_BaseDataLoaderIter` object every epoch,
In the case of the multiprocess DataLoader, every epoch the worker processes are re-created and they make a copy of the original `Dataset` object.
If users want to cache data or do some tracking on their datasets, all their data will be wiped out every epoch. Notice that this doesn't happen when the number of workers is 0. giving some inconsistencies with the multiprocess and serial data loaders.

This PR keeps the `_BaseDataLoaderIter` object alive and just resets it within epochs, so the workers remain active and so their own `Dataset` objects. People seem to file issues about this often.

Commit:
Reset `_BaseDataLoaderIter` instead of creating a new one

ezyang commented:
@SsnL do you think you could take a look at this? Lmk if you can't and I'll find someone else.

tkerola reviewed (commented):
I just added some small comments.

SsnL commented:
Thanks for doing this. Some of the issues that we need to figure out for worker reusing are 
1. whether we should enable this by default, as workers may hold resources.
2. what about seeds and `worker_init_fn`? do we reset / run them at each epoch?

Honestly speaking I don't know the correct answer to these two questions. What are your thoughts? :)

emcastillo commented on torch/utils/data/dataloader.py:
```diff
@@ -335,6 +337,10 @@ def __init__(self, loader):
     def __iter__(self):
         return self
 
+    def _reset(self):
```
Thanks!

emcastillo commented:
Hi! Thanks for the feedback 😃

> whether we should enable this by default, as workers may hold resources.


I think this is the expected behavior by users so it's not a bad idea to have it as the default.
However, this may cause some no-compat issues with users code so we have to be careful here, maybe adding a `continuous` parameter to the constructor is good?

> what about seeds and worker_init_fn? do we reset / run them at each epoch?


I think that running then only once should be fine.

emcastillo commented:
I added an argument to `DataLoader` constructor to control the behavior after discussing it with @tkerola 
Fixed some issues and I guess that all that is left is to write some unit tests.

emcastillo commented:
Hi, @SsnL can we get some comments of this? 😇

emcastillo commented:
Hello,
We would like to get some comments if possible!
Thanks

SsnL commented:
Hi sorry for the lateness. I will take a look by tomorrow.

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -328,13 +342,24 @@ def __init__(self, loader):
         self._base_seed = torch.empty((), dtype=torch.int64).random_().item()
-        self._num_yielded = 0
+        self._persistent_workers = loader._persistent_workers
+        self._reset_base()
```
This is hacky. Can we eliminate `_reset_base` instead call `self._reset()` in `__iter__` before `return self`?

emcastillo replied:
Sure, let me give it a try :)

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -233,6 +237,16 @@ def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None,
         self._IterableDataset_len_called = None  # See NOTE [ IterableDataset and __len__ ]
 
+        self._persistent_workers = persistent_workers
+        self._iterator = self._get_iterator() if persistent_workers else None
```
is it intended that the workers are started when `DataLoader` is created (rather than when the first iterator of `DataLoader` is created)?

ssnl reviewed (commented):
Also need to reset the dataset iter in worker process for IterableDataset

SsnL commented:
Thanks again for doing this! My general opinion is that this should definitely be done. But given how complicated the current DataLoader ctor arguments are (look at the # of arguments that are only used with multiprocessing loading), and equally unhappy about how complicated the DataLoader logic is, we should be careful about the behavior and API.

emcastillo commented:
Hi, Thanks a lot for the comments!

Regarding resetting the `IterableDataset` iterator in the worker, I can't find a reset mechanism for it.
https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#IterableDataset
Maybe I am looking at the wrong place?

SsnL commented:
@emcastillo It would require recreating the iterator in the worker process, after all iterators exhaust. So some kind of message passing from the main process to the worker is needed for this. You can probably alter the index_queue mechanism to do so.

emcastillo commented:
Hi, I did the synchronization of the worker process and the iterator itself.

Regarding the `DataLoader` API, the only alternative I see is to create a sub-class rather than adding a parameter. But since there are no other specializations I find it a bit overkill.

emcastillo commented:
I have changed the dataloader tests in my local env to use always the persistent workers.
All works but one test where a warning after accessing length is not being raised.

I still have to deal with it, but all the other tests work fine.
Once I fix this, I will push a specialization of the test suite for the persistent workers version.
I am just thinking about creating a factory function for the `DataLoader` so I can subclass the `TestDataLoader` and use this function instead of adding or modifying all the existing tests.

emcastillo commented:
@ssnl any comments?😇

emcastillo commented:
Just dropping by 😁

emcastillo commented:
Hi @SsnL , I rebased and solved conflicts, can I get another look?
Thanks!

ssnl reviewed (commented):
Generally great patch. Thanks for working on this and sorry for the late reply. I had a few comments. Let me know what you think!

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -123,7 +126,8 @@ def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None,
                  worker_init_fn=None, multiprocessing_context=None,
-                 generator=None):
+                 generator=None,
+                 persistent_workers=False):
```
Also let's make this keyword only and check that users don't set `num_workers > 0` and `persistent_worker=True` :)

emcastillo commented on test/test_dataloader.py:
```diff
@@ -1080,20 +1090,21 @@ def test_iterable_style_dataset(self):
+                    self.assertEqual(w.exitcode, 0)
+            finally:
+                for w in workers:
+                    w.terminate()
```
actually this checks works for both, I don't know why I did this ...

emcastillo commented:
Hey @SsnL 
Thanks a lot for the great feedback :)
I think I addressed all your points, May I get another pass?
We would like this to be in 1.6 if possible 😇

SsnL commented:
@emcastillo CI error seems real, and easily fixable :)

SsnL commented:
also cc @VitalyFedyunin

emcastillo commented:
@SsnL think tests should be ok now!

thanks a lot ☺️

SsnL commented on test/test_dataloader.py:
```diff
@@ -1128,7 +1143,7 @@ def test_iterable_style_dataset(self):
         # [auto-batching] test that workers exit gracefully
         workers = dataloader_iter._workers
-        del dataloader_iter
+        del dataloader
```
you probably need to `del` both since both have references to iter

SsnL commented on test/test_dataloader.py:
```diff
@@ -1164,7 +1179,7 @@ def test_iterable_style_dataset(self):
         # [auto-batching & drop_last] test that workers exit gracefully
         workers = dataloader_iter._workers
-        del dataloader_iter
+        del dataloader
```
similar here

SsnL commented on test/test_dataloader.py:
```diff
@@ -752,6 +756,17 @@ def setUp(self):
+                ('persistent_workers' in kwargs and kwargs['persistent_workers'])):
+            if (('num_workers' in kwargs and kwargs['num_workers'] == 0)
+                    or ('num_workers' not in kwargs)):
+                self.skipTest('persistent_workers and num_workers == 0 is not compatible')
```
I understand where this is coming from, but this skips a bunch of tests where only part of them uses num_workers=0. Could you instead turn off `persistent_worker=False` when `num_workers=0` rather than skipping?

SsnL commented:
Also remember to update the signature here
https://github.com/pytorch/pytorch/blob/67a4f375cdf06113ca959b4e16739edb666f243f/docs/source/data.rst#L22-L25

emcastillo commented:
I think it should be ok now 😬

SsnL commented on test/test_dataloader.py:
```diff
@@ -759,13 +759,16 @@ def setUp(self):
+        else:
+            persistent_workers = self.persistent_workers
+        if persistent_workers :
             if (('num_workers' in kwargs and kwargs['num_workers'] == 0)
```
:) one minor style recommendation: test `if kwargs.get('num_workers', 0) == 0`.

also just to be safe, can we only allow a single `dataset` as the positional argument, i.e., `def _get_data_loader(self, dataset, **kwargs):`? This guarantees that we can test `kwargs` for things like `num_workers`.

emcastillo replied:
Sorry I overlooked this, let me fix it :)

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -1105,8 +1105,6 @@ def _mark_worker_as_unavailable(self, worker_id, shutdown=False):
         # this method will lead to the worker process ending if _workers_done_event
         # is set, but will only mark it as unavailable for this epoch otherwise.
 
-        assert self._workers_status[worker_id]
```
Hmm, when would a worker be marked twice?

emcastillo replied:
The worker can be marked twice in the case of `persistent_workers=True`.
First, if the StopIteration has happened, the worker will be marked unavailable, but it won't be finished.

Then, on the shutdown event, the worker will be paused, but we need it to exit the loop, so we send the `None` event even if it was paused.

I will try to improve the checks here and add comments. Maybe the status should not be changed for persistent workers until the real shutdown?

I agree this is not intuitive and clean.

SsnL replied:
First, status should definitely be changed, otherwise the subsequent keys won't be sent to correct workers.

How about this:
1. Let's keep the original invariance: sending `None` means that the worker will be terminated. 
2. So, in the case of `persistent_workers`, and receiving `_IterableDatasetStopIteration`, we simply just mark that worker's `worker_status` to be `False`, *without* calling `_shutdown_worker` (or `_mark_worker_as_unavailable` with the new name). 
3. In worker loop, we also don't do anything special if we receive `None` and `persistent_worker`. 

I personally feel that this way the code will be much simpler, and we won't need to change anything related to `_shutdown_worker`...

SsnL replied:
Let me know what you think!

emcastillo replied:
Sounds good!! 
I think my latest commit is in line of that, I will just set explicitly the status to `False` when `_IterableDatasetStopIteration` happens and I think that should be enough.

emcastillo replied:
Done, PTAL :)

VitalyFedyunin commented:
Hi! I'm still checking implementation, but it would be nice to see an actual test which checks that self.persistent_workers = True doesn't reset DataSet. 

Also please correct me if I'm wrong, but we just doubled number of DataLoader tests =)

emcastillo commented:
> Also please correct me if I'm wrong, but we just doubled number of DataLoader tests =)

Yes!, I felt lazy so I just re-run the tests with the `persistent_workers` parameter set to `True` 😅

I will also try to make a better shutdown mechanism, current one is a bit fragile.

emcastillo commented:
Hi @VitalyFedyunin , I just added a simple test to verify that the Dataset is not being recreated every epoch

VitalyFedyunin commented:
Thanks! I plan to review this PR in details. But generally I like idea and reset instead of respawn was one of the goals for 1.7-1.8. 

Please take a look into Windows build when you have a chance.

emcastillo commented on test/test_dataloader.py:
```diff
@@ -1923,6 +1923,38 @@ def __next__(self):
+                return self.data[idx]
+
+        dataset = DummyDataset()
+        dataloader = self._get_data_loader(dataset, num_workers=2)
```
I think this is an outdated commit, I fixed these yesterday

SsnL replied:
Oh sorry! Didn't notice!

emcastillo replied:
no prob 😄

emcastillo commented:
Any feedback in this 😇

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -787,7 +820,28 @@ def __init__(self, loader):
+        self._workers_status = [True for i in range(self._num_workers)]
+        # We resume the prefetching in case it was enabled
+        if not first_iter:
+            # 
```
empty comment :)

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -1043,12 +1100,16 @@ def _process_data(self, data):
 
-        assert self._workers_status[worker_id]
+        # this method will lead to the worker process ending if _workers_done_event
+        # is set, but will only mark it as unavailable for this epoch otherwise.
```
I don't think that this is true anymore.  Even when the case of `persistent_worker = True`, this is still called just once per worker, right?

emcastillo replied:
Exactly, this is called just once per worker, this is a stale comment, thanks a lot for the careful review 🙇‍♂️

ssnl reviewed (commented):
Looks mostly good to me!  also cc @VitalyFedyunin

VitalyFedyunin commented:
Hello! I'm allocating a full day tomorrow to review and land it. Meanwhile can you please rebase.

VitalyFedyunin commented on test/test_dataloader.py:
```diff
@@ -1846,6 +1871,85 @@ def test_pin_memory(self):
+            # Changing the start value here doesn't have any effect in the dataset
+            # cached by the workers. since they are not recreated between epochs
+            # and can cache values safely
+            dataset.start = i
```
It is failing nicely with `self.persistent_workers = False` ?

VitalyFedyunin replied:
It is confirmed to fail (which is expected), but that it is actually checking is the fact that we are not making any new copies of `dataset`. There is no test which checks that you are using exactly same DataSet objects after reset. 

It is still acceptable to merge as is anyway.

emcastillo replied:
Actually, the `DataSet` object is instantiated only once, by the user, and then it seems that is cloned when calling:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L764-L769

The object seems to be hardly cloned in the new process memory space, no pickle (I tried to verify this by overriding `__getstate__` or `__setstate__`, additionally, in the new processes, the object id is the same than in the main process, so there is no actual "instantiation".

It seems that the only way we have to assert that the object is not being recreated is actually this test. If the object is recreated it will get new values in the next iteration (.start) and if not, it will retain the cached values.

VitalyFedyunin commented on torch/utils/data/dataloader.py:
```diff
@@ -743,6 +772,9 @@ def __init__(self, loader):
+        # Not that this indicates that a worker still has work to do *for this epoch*.
+        # It does not mean that a worker is dead. In case of `_persistent_workers`, 
+        # the worker will be reset to available in the next epoch.
         self._workers_status = []
```
nit: This going to be overwritten by following `_reset` call

VitalyFedyunin reviewed (commented):
Overall looks good with couple concerns:

1) I'm not sure we are actually testing that we reset DataSet instead of creating new ones. We might need to create another multiprocessing queue to gather number of initializations and asserting on them.

2) There is no way for DataSet to tell if it is used in classic or reset mode. If might be error-prone if DataSet tries to free memory, especially if we do iteratable datasets

emcastillo commented:
Hi!, Thanks for all the reviews and feedback,
I am a bit busy with some internal projects right now, so once I get everything sorted up I will rebase and change this accordingly.

Thanks again!

emcastillo commented:
Sorry for the long time without taking care of this.
I just rebased and verified that everything works fine, I will be addressing the concerns in the review in the next hours.

emcastillo commented:
Hi, I tried to do what you suggest but I had a bit of hard time:

> I'm not sure we are actually testing that we reset DataSet instead of creating new ones. We might need to create another multiprocessing queue to gather number of initializations and asserting on them.

As I replied above, I tried to do similar things to enable this check, but it seems that the object is being directly cloned in the fork without any actual python instantiation or pickling (it always retain the original object id). I also verified that `__new__` is also called once, in the main process.
It seems that we don't have an easy way to assert that the object is the same or not.

> There is no way for DataSet to tell if it is used in classic or reset mode. If might be error-prone if DataSet tries to free memory, especially if we do iterable datasets.

Since the persistent workers are not enabled by default, I believe up to a certain extent that it is the user responsibility to ensure that their datasets can be used in this manner. We could add some kind of facility to the Dataset to identify the mode, (probably an attribute with a setter method invoked from the dataloader?) but I think we can leave that to a different PR if possible.

emcastillo commented:
Thanks for approving the changes!!
Will this be landed soon?😇

VitalyFedyunin commented:
Sorry it takes eternity to land. Changes to DataLoader triggered all ML tests to run.

SsnL commented:
Glad to see this merged. Thanks @emcastillo and @VitalyFedyunin !
